Document teleport-operator (#14965)
gravitational/teleport: docs/pages/setup/guides/teleport-operator.mdx

---
title: Teleport Kubernetes Operator
description: Easily manage Teleport resources from Kubernetes
---

> The Teleport Kubernetes Operator provides a way for Kubernetes users to manage Teleport resources like Users and

Author: I find it weird to not capitalize kinds as we are following the same resource/kind pattern than Kubernetes, I would have expected us to follow their convention of writing kinds capitalized.

Roles using Kubernetes resources.

<Admonition type="note">
  Setting up the Teleport Operator is not required to
  [deploy Teleport on Kubernetes](../helm-deployments.mdx), nor setup
  [Kubernetes Access](../../kubernetes-access/introduction.mdx).
</Admonition>

This guide covers how to:
- Install Teleport Operator on Kubernetes.
- Configure Teleport users and roles using Kubernetes.

## Prerequisites

(!docs/pages/includes/edition-prereqs-tabs.mdx!)

- Kubernetes cluster that will host Teleport;
- [Helm](https://helm.sh/docs/intro/quickstart/)
- [kubectl](https://kubernetes.io/docs/tasks/tools/)

<Admonition type="tip">
  Users wanting to experiment locally with the operator can use [minikube](https://minikube.sigs.k8s.io/docs/start/)
  to start a local Kubernetes cluster:
  ```code
  $ minikube start
  ```
</Admonition>

> ## Step 1/3. Ensure Kubernetes cluster is running

Author: Done

$ kubectl exec -it "$PROXY_POD" -c teleport -- tctl users ls
# User                          Roles
# ----------------------------- -----------------------------
# bot-teleport-operator-sidecar bot-teleport-operator-sidecar
# myuser                        myrole
```

At this point the Teleport Kubernetes Operator is functional and Teleport users and roles can be managed from
Kubernetes.

## Going further

### Exploring Teleport Kubernetes CRDs

Available fields can be browsed with `kubectl explain` in a cluster with Teleport CRDs installed.
For example the command:
```code
$ kubectl explain teleportroles.spec
```
Returns the following fields:
```shell
KIND:     TeleportRole
VERSION:  resources.teleport.dev/v5

RESOURCE: spec <Object>

DESCRIPTION:
    Role resource definition v5 from Teleport

FIELDS:
   allow	<Object>
     Allow is the set of conditions evaluated to grant access.

   deny	<Object>
     Deny is the set of conditions evaluated to deny access. Deny takes priority
     over allow.

   options	<Object>
     Options is for OpenSSH options like agent forwarding.
```

### Troubleshooting a resource not reconciled in Teleport

> During reconciliation the operator adds a `status` field on the Kubernetes resource.

Reviewer: I think we should explain reconciliation and the overall lifecycle of resources managed by the Operator. Right now, this is the first time the guide mentions reconciliation, and not all users may be familiar with the operator pattern.

> If Teleport returns some error it will be available in `status.conditions`.

Reviewer: ```suggestion
If Teleport returns an error it will be available in `status.conditions` when you retrieve a resource via `kubectl`:
```

```code
$ kubectl get teleportusers myuser -o yaml
```

> For example, if a user has been granted a non-exising role the status will look like:

Reviewer: ```suggestion
For example, if a user has been granted a nonexistent role the status will look like:
```

```yaml
apiVersion: resources.teleport.dev/v2
kind: TeleportUser
# [...]
status:
  conditions:
  - lastTransitionTime: "2022-07-25T16:15:52Z"
    message: Teleport resource has the Kubernetes origin label.
    reason: OriginLabelMatching
    status: "True"
    type: TeleportResourceOwned
  - lastTransitionTime: "2022-07-25T17:08:58Z"
    message: 'Teleport returned the error: role my-non-existing-role is not found'
    reason: TeleportError
    status: "False"
    type: SuccessfullyReconciled
```

Here `SuccessfullyReconciled` is `False` and the error is `role my-non-existing-role is not found`.

> If the status is not present or does not gives sufficient information to solve the issue, check the operator logs:

Reviewer: ```suggestion
If the status is not present or does not give sufficient information to solve the issue, check the operator logs:
```

```shell
$ kubectl logs "$PROXY_POD" -c operator
```

<Admonition type="note">

>   In case of multi-replicas deployments, only one operator instance is active.

Reviewer: ```suggestion
  In case of multi-replica deployments, only one operator instance is active.
```

>   In case of multi-replicas deployments, only one operator instance is active.

Reviewer: What does "active" mean here? Also, this is the first time we mention leader election—there should at least be some explanation. And will the user need to check the logs of all Teleport Operator pods?

Author: Rephrased

---
title: Teleport Kubernetes Operator (Preview)
description: Easily manage Teleport resources from Kubernetes
---

<Admonition type="warning" title="Preview">
  Teleport Kubernetes Operator is currently in Preview mode.

> </Admonition>

Reviewer 2: Nit: I don't know if we need to include this admonition if the title already says "preview". I'm not sure we've done this for other preview features since introducing the new product preview policy.

@ptgott @xinding33 Not sure if you folks have an opinion on this.

Author: I copied this from another preview page: pages/database-access/guides/sql-server-ad.mdx

The Teleport Kubernetes Operator provides a way for Kubernetes users to manage some Teleport resources through
Kubernetes, following [the Operator Pattern](https://kubernetes.io/docs/concepts/extend-kubernetes/operator/).

The Teleport Kubernetes Operator is deployed alongside its CustomResourceDefinitions. Once deployed, users
can use a Kubernetes client like `kubectl` or their existing CI/CD Kubernetes pipelines to create Teleport
CustomResources. The Teleport Kubernetes Operator watches for those resources and does API calls to Teleport to
reach the desired state.

Currently supported Teleport resources are:
- users
- roles

This guide covers how to:
- Install Teleport Operator on Kubernetes.

> - Configure Teleport users and roles using Kubernetes.

Reviewer 2: ```suggestion
- Manage Teleport users and roles using Kubernetes.
```

## Prerequisites

(!docs/pages/includes/edition-prereqs-tabs.mdx!)

> - Kubernetes cluster (with or without `teleport-cluster` already deployed);

Reviewer 2: I would say "teleport-cluster Helm chart", otherwise not clear what teleport-cluster is referring to for new users.

- [Helm](https://helm.sh/docs/intro/quickstart/)
- [kubectl](https://kubernetes.io/docs/tasks/tools/)

Validate Kubernetes connectivity by running the following command:

```code
$ kubectl cluster-info
# Kubernetes control plane is running at https://127.0.0.1:6443
# CoreDNS is running at https://127.0.0.1:6443/api/v1/namespaces/kube-system/services/kube-dns:dns/proxy
# Metrics-server is running at https://127.0.0.1:6443/api/v1/namespaces/kube-system/services/https:metrics-server:https/proxy
```

<Admonition type="tip">
  Users wanting to experiment locally with the operator can use [minikube](https://minikube.sigs.k8s.io/docs/start/)
  to start a local Kubernetes cluster:
  ```code
  $ minikube start
  ```
</Admonition>

> ## Step 1/3. Install teleport-cluster with the operator

Reviewer 2: ```suggestion
## Step 1/3. Install teleport-cluster Helm chart with the operator
```

Add the Teleport Helm Chart repo:
```code

> $ helm repo add teleport https://charts.releases.teleport.dev

Reviewer 2: Nit: May wanna instruct users to run `helm repo update` also in case they already have the repo.

```

Install the Helm chart for the Teleport Cluster with `operator.enabled=true`.

```code
$ helm install teleport-cluster teleport/teleport-cluster \
        --create-namespace --namespace teleport-cluster \

>         --set clusterName=teleport-cluster.teleport-cluster.svc.cluster.local \

Reviewer 2: Nit: Can't use something simpler, like `example`, as a cluster name?

        --set operator.enabled=true \
        --set teleportVersionOverride=(=teleport.version=)
```

This command installs the required Kubernetes CRDs and deploys the Teleport Kubernetes Operator next to the Teleport
cluster. All resources (except the CRDs which are cluster-scoped) are created in the `teleport-cluster` namespace.

All subsequents commands will take place in the `teleport-cluster` namespace, so let's use it by default:
```code
$ kubectl config set-context --current --namespace teleport-cluster
```

> ## Step 2/3. Manage users and roles using `kubectl`

Reviewer 2: ```suggestion
## Step 2/3. Manage Teleport users and roles using `kubectl`
```

```code
$ PROXY_POD=$(kubectl get po -l app=teleport-cluster -o jsonpath='{.items[0].metadata.name}')
$ kubectl exec -it "$PROXY_POD" -c teleport -- tctl users ls
# User                          Roles
# ----------------------------- -----------------------------
# bot-teleport-operator-sidecar bot-teleport-operator-sidecar
# myuser                        myrole
```

At this point the Teleport Kubernetes Operator is functional and Teleport users and roles can be managed from
Kubernetes.

## Step 3/3. Explore the Teleport CRDs

Available fields can be browsed with `kubectl explain` in a cluster with Teleport CRDs installed.
For example the command:
```code
$ kubectl explain teleportroles.spec
```
Returns the following fields:
```shell
KIND:     TeleportRole
VERSION:  resources.teleport.dev/v5

RESOURCE: spec <Object>

DESCRIPTION:
    Role resource definition v5 from Teleport

FIELDS:
   allow	<Object>
     Allow is the set of conditions evaluated to grant access.

   deny	<Object>
     Deny is the set of conditions evaluated to deny access. Deny takes priority
     over allow.

   options	<Object>
     Options is for OpenSSH options like agent forwarding.
```

> ## Going Further

Reviewer 2: ```suggestion
## Troubleshooting
```

> ### Troubleshooting a resource not reconciled in Teleport

Reviewer 2: ```suggestion
### Resource is not reconciled in Teleport
```

  conditions:
  - lastTransitionTime: "2022-07-25T16:15:52Z"
    message: Teleport resource has the Kubernetes origin label.
    reason: OriginLabelMatching
    status: "True"
    type: TeleportResourceOwned
  - lastTransitionTime: "2022-07-25T17:08:58Z"
    message: 'Teleport returned the error: role my-non-existing-role is not found'
    reason: TeleportError
    status: "False"
    type: SuccessfullyReconciled
```

Here `SuccessfullyReconciled` is `False` and the error is `role my-non-existing-role is not found`.

If the status is not present or does not give sufficient information to solve the issue, check the operator logs:

```shell
$ kubectl logs "$PROXY_POD" -c operator
```

<Admonition type="note">
  In case of multi-replica deployments, only one operator instance is running the reconciliation loop. This operator
  is called the leader and is the only one producing reconciliation logs. The other operator instances are waiting with
  the following log:

  ```
  leaderelection.go:248] attempting to acquire leader lease teleport/431e83f4.teleport.dev...
  ```

  To diagnose reconciliation issues, you will have to inspect all pods to find the one reconciling the resources.

> </Admonition>

Reviewer 2: Nit: Add a "next steps" section at the end with a few links to e.g. Helm chart reference, access controls and [deployment guides](https://goteleport.com/docs/setup/helm-deployments/).

See other guides for examples, e.g. database access.

Author: Done

---
title: Teleport Kubernetes Operator (Preview)
description: Easily manage Teleport resources from Kubernetes
---

<Admonition type="warning" title="Preview">
  Teleport Kubernetes Operator is currently in Preview mode.
</Admonition>

The Teleport Kubernetes Operator provides a way for Kubernetes users to manage some Teleport resources through
Kubernetes, following [the Operator Pattern](https://kubernetes.io/docs/concepts/extend-kubernetes/operator/).

> The Teleport Kubernetes Operator is deployed alongside its CustomResourceDefinitions. Once deployed, users

Reviewer: ```suggestion
The Teleport Kubernetes Operator is deployed alongside its custom resource definitions. Once deployed, users
```
Using the convention in the Kubernetes docs: https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/

can use a Kubernetes client like `kubectl` or their existing CI/CD Kubernetes pipelines to create Teleport

> CustomResources. The Teleport Kubernetes Operator watches for those resources and does API calls to Teleport to

Reviewer: ```suggestion
custom resources. The Teleport Kubernetes Operator watches for those resources and does API calls to Teleport to
```

        --set teleportVersionOverride=(=teleport.version=)
```

This command installs the required Kubernetes CRDs and deploys the Teleport Kubernetes Operator next to the Teleport

> cluster. All resources (except the CRDs which are cluster-scoped) are created in the `teleport-cluster` namespace.

Reviewer: ```suggestion
cluster. All resources (except CRDs, which are cluster-scoped) are created in the `teleport-cluster` namespace.
```
Suggested to make it clearer that all CRDs are cluster scoped. If not, we could change to "except the CRDs that are cluster scoped".

All subsequents commands will take place in the `teleport-cluster` namespace, so let's use it by default:
```code
$ kubectl config set-context --current --namespace teleport-cluster
```

## Step 2/3. Manage users and roles using `kubectl`

> Create a manifest `teleport-resources.yaml` that describes two CustomResources: a `TeleportUser` and a `TeleportRole`:

Reviewer: ```suggestion
Create a manifest called `teleport-resources.yaml` that describes two custom resources: a `TeleportUser` and a `TeleportRole`:
```

     Options is for OpenSSH options like agent forwarding.
```

## Going Further

### Troubleshooting a resource not reconciled in Teleport

> The operator watches for new resources or changes in Kubernetes. When a change happens, it triggers the reconciliation

Reviewer: ```suggestion
The Teleport Operator watches for new resources or changes in Kubernetes. When a change happens, it triggers the reconciliation
```
